lib/greengrass.ts

```diff
@@ -0,0 +1,157 @@
+    }
+
+    static from_json(json: any) {
+        const cores: GGCore[] = [];
+        if (json.Cores && isArray(json.Cores)) {
+            json.Cores.forEach((payload: any) => {
+                cores.push(GGCore.from_json(payload));
+            });
+        }
+        return new GGGroup(
+            json.GGGroupId,
+            cores,
+            json.CAs
+        )
+    }
+}
+
+export class DiscoverResponse {
+    private constructor(
+        readonly gg_groups: GGGroup[] = []) {
+        
+    }
+
+    static from_json(json: any) {
+        const groups: GGGroup[] = [];
+        if (json.GGGroups && isArray(json.GGGroups)) {
+            json.GGGroups.forEach((payload: any) => {
+                groups.push(GGGroup.from_json(payload));
+            });
+        }
+        return new DiscoverResponse(groups);
+    }
+}
+
+export class DiscoveryClient {
+    private connection_manager: http.HttpClientConnectionManager;
+    private endpoint: string;
+    constructor(
+        readonly bootstrap: io.ClientBootstrap,
+        private socket_options: io.SocketOptions,
+        private tls_ctx: io.ClientTlsContext,
+        readonly region: string
+    ) {
+        this.endpoint = `greengrass-ats.iot.${region}.amazonaws.com`;
+        this.connection_manager = new http.HttpClientConnectionManager(
+            this.bootstrap,
+            this.endpoint,
+            io.is_alpn_available() ? 443: 8443,
+            4,
+            16 * 1024,
+            this.socket_options,
+            this.tls_ctx
+        );
+    }
+
+    discover(thing_name: string) : Promise<DiscoverResponse> {
+        return new Promise(async (resolve, reject) => {
+            this.connection_manager.acquire()
+                .then((connection) => {
+                    const request = new http.HttpRequest(
+                        'GET', `/greengrass/discover/thing/${thing_name}`,
+                        undefined, new http.HttpHeaders([['host', this.endpoint]]));
+                    const stream = connection.request(request);
+                    let response = '';
+                    const decoder = new TextDecoder('utf8');
+                    stream.on('response', (status_code, headers) => {
+                        if (status_code != 200) {
+                            reject(new DiscoveryError(`Discovery failed (headers: ${headers})`, status_code));
+                        }
+                    });
+                    stream.on('data', (body_data) => {
+                        response += decoder.decode(body_data);
+                    });
+                    stream.on('end', () => {
+                        const json = JSON.parse(response);
+                        const discover_response = DiscoverResponse.from_json(json);
+                        resolve(discover_response);
+                    });
+                    stream.on('error', (error) => {
+                        reject(new DiscoveryError(error.toString()));
+                    });
```

Comment:
Are JS programmers supposed to subscribe to 'error' events first to prevent timing issues where 'response' uncorks the emitter, but we may not be subscribed to 'error' yet so a throw might happen?

Reply:
I have no idea, I'd expect so? I suspect that's why it throws if no one is subscribed, so you can't miss it. Also, keep in mind that the uncork is not synchronous, so if all your ```.on()``` calls are in a row, the events are delivered next tick.

Reply:
ah. I like that magic sauce. Thanks node